src/anyio/_backends/_asyncio.py

```diff
@@ -1052,12 +1052,26 @@ async def aclose(self) -> None:
 @dataclass(eq=False)
 class StreamWriterWrapper(abc.ByteSendStream):
     _stream: asyncio.StreamWriter
+    _closed: bool = field(init=False, default=False)
 
     async def send(self, item: bytes) -> None:
-        self._stream.write(item)
-        await self._stream.drain()
+        await AsyncIOBackend.checkpoint()
+        try:
+            self._stream.write(item)
+            await self._stream.drain()
+        except (ConnectionResetError, BrokenPipeError, RuntimeError) as exc:
+            # If closed by us and/or the peer:
+            # * on stdlib, drain() raises ConnectionResetError or BrokenPipeError
+            # * on uvloop, write() raises RuntimeError. drain() probably raises other
+            #   things too, but it's hard to test.
+            if self._closed:
+                raise ClosedResourceError from exc.__cause__
+            elif self._stream.is_closing():
+                raise BrokenResourceError from exc.__cause__
+            raise
```

Comment:
We could also use a custom `Protocol.connection_lost` instead of catching/remapping these exceptions, i.e. use `BaseEventLoop.subprocess_exec` instead of the higher-level `asyncio.create_subprocess_exec`. That seems a bit annoying to do however, and it would just be deleted if/when we pursue #783 anyway. In the meantime, catching the exceptions seems to be an okay fix...it's possible that we have missed some exception types in the `except` tuple, but it seems to me that the exception semantics of this function should be no worse than they are without this exception remapping.

Reply:
Why are we chaining these from `exc.__cause__` and not `exc` directly?

Reply:
Fixed. I had incorrectly borrowed that from

```python
except trio.ClosedResourceError as exc:
    raise ClosedResourceError from exc.__cause__
```